File: algorithms/src/Graph/DirectedGraph.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;

namespace Algorithms.Graph
{
    public class DirectedGraph<TVertex, TEdge> :
        IEnumerable<TVertex>,
        IEquatable<DirectedGraph<TVertex, TEdge>>,
        ICloneable
    {
        struct VertexData: IEquatable<VertexData>
        {
            public Dictionary<TVertex, TEdge> Outs;
            public Dictionary<TVertex, TEdge> Ins;

            public static VertexData Default()
            {
                return new VertexData()
                {
                    Outs = new Dictionary<TVertex, TEdge>(),
                    Ins = new Dictionary<TVertex, TEdge>(),
                };
            }

            public bool Equals(VertexData other)
            {
                bool @out = this.Outs.All((KeyValuePair<TVertex, TEdge> pair) =>
                {
                    if (!other.Outs.ContainsKey(pair.Key))
                    {
                        return false;
                    }

                    return pair.Value.Equals(other.Outs[pair.Key]);
                });

                bool @in = this.Ins.All((KeyValuePair<TVertex, TEdge> pair) =>
                {
                    if (!other.Ins.ContainsKey(pair.Key))
                    {
                        return false;
                    }

                    return pair.Value.Equals(other.Ins[pair.Key]);
                });

                return @out && @in;
            }
        }

        Dictionary<TVertex, VertexData> _dictionary = new Dictionary<TVertex, VertexData>();

        /// <summary>
        /// Get a number of vertices
        /// </summary>
        public int VertexCount => _dictionary.Count;

        public void Add(TVertex vertex)
        {
            _dictionary.Add(vertex, VertexData.Default());
        }

        public void Add(TVertex @out, TVertex @in, TEdge e)
        {
            VertexData fromVertex = _dictionary[@out];
            VertexData inVertex = _dictionary[@in];

            fromVertex.Outs.Add(@in, e);
            inVertex.Ins.Add(@out, e);

            _dictionary[@out] = fromVertex;
            _dictionary[@in] = inVertex;
        }

        public Dictionary<TVertex, TEdge> GetOuts(TVertex vertex)
        {
            return _dictionary[vertex].Outs;
        }

        public Dictionary<TVertex, TEdge> GetIns(TVertex vertex)
        {
            return _dictionary[vertex].Ins;
        }

        public bool Contains(TVertex vertex)
        {
            return _dictionary.ContainsKey(vertex);
        }

        public IEnumerator<TVertex> GetEnumerator()
        {
            foreach (var pair in _dictionary)
            {
                yield return pair.Key;
            }
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return this.GetEnumerator();
        }

        public bool Equals(DirectedGraph<TVertex, TEdge> other)
        {
            return _dictionary.All((KeyValuePair<TVertex, VertexData> pair) =>
            {
                if (!other._dictionary.ContainsKey(pair.Key))
                {
                    return false;
                }

                return pair.Value.Equals(other._dictionary[pair.Key]);
            });
        }

        public object Clone()
        {
            var copy = new DirectedGraph<TVertex, TEdge>();

            foreach (var pair in _dictionary)
            {
                var outs = new Dictionary<TVertex, TEdge>();
                var ins = new Dictionary<TVertex, TEdge>();

                foreach (var @out in pair.Value.Outs)
                {
                    outs.Add(@out.Key, @out.Value);
                }

                foreach (var @in in pair.Value.Ins)
                {
                    ins.Add(@in.Key, @in.Value);
                }

                copy._dictionary.Add(pair.Key, new VertexData()
                {
                    Outs = outs,
                    Ins = ins,
                });
            }

            return copy;
        }
    }
}

File: python/algorithms/graph/directed_graph.py
from typing import TypeVar, Generic, Dict, Iterator
from dataclasses import dataclass, field

TVertex = TypeVar('TVertex')
TEdge = TypeVar('TEdge')


@dataclass
class VertexData(Generic[TVertex, TEdge]):
    """Data structure to hold outgoing and incoming edges for a vertex."""
    outs: Dict[TVertex, TEdge] = field(default_factory=dict)
    ins: Dict[TVertex, TEdge] = field(default_factory=dict)


class DirectedGraph(Generic[TVertex, TEdge]):
    """A directed graph implementation with generic vertex and edge types."""

    def __init__(self):
        self._dictionary: Dict[TVertex, VertexData[TVertex, TEdge]] = {}

    @property
    def vertex_count(self) -> int:
        """Get the number of vertices in the graph."""
        return len(self._dictionary)

    def add_vertex(self, vertex: TVertex) -> None:
        """Add a vertex to the graph."""
        if vertex not in self._dictionary:
            self._dictionary[vertex] = VertexData()

    def add_edge(self, from_vertex: TVertex, to_vertex: TVertex, edge: TEdge) -> None:
        """Add an edge between two vertices."""
        if from_vertex not in self._dictionary:
            self.add_vertex(from_vertex)
        if to_vertex not in self._dictionary:
            self.add_vertex(to_vertex)

        self._dictionary[from_vertex].outs[to_vertex] = edge
        self._dictionary[to_vertex].ins[from_vertex] = edge

    def get_outs(self, vertex: TVertex) -> Dict[TVertex, TEdge]:
        """Get all outgoing edges from a vertex."""
        return self._dictionary[vertex].outs

    def get_ins(self, vertex: TVertex) -> Dict[TVertex, TEdge]:
        """Get all incoming edges to a vertex."""
        return self._dictionary[vertex].ins

    def contains(self, vertex: TVertex) -> bool:
        """Check if the graph contains a vertex."""
        return vertex in self._dictionary

    def __iter__(self) -> Iterator[TVertex]:
        """Iterate over all vertices in the graph."""
        return iter(self._dictionary.keys())

    def __eq__(self, other) -> bool:
        """Check equality with another DirectedGraph."""
        if not isinstance(other, DirectedGraph):
            return False

        if len(self._dictionary) != len(other._dictionary):
            return False

        for vertex, data in self._dictionary.items():
            if vertex not in other._dictionary:
                return False

            other_data = other._dictionary[vertex]
            if data.outs != other_data.outs or data.ins != other_data.ins:
                return False

        return True

    def clone(self) -> 'DirectedGraph[TVertex, TEdge]':
        """Create a deep copy of the graph."""
        cloned = DirectedGraph[TVertex, TEdge]()

        for vertex, data in self._dictionary.items():
            cloned._dictionary[vertex] = VertexData(
                outs=data.outs.copy(),
                ins=data.ins.copy()
            )

        return cloned
